### PR TITLE
KAFKA-12248: Add BrokerHeartbeat/BrokerRegistration RPCs for KIP-500

### PR DESCRIPTION
**Pull request by aloknnikhil**
### Committer Checklist (excluded from commit message)
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status
- [x] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by hachikuji** [comment]
Thanks for the PR. Just to make this a little more manageable, could we split this into two PRs? One for the register/heartbeat controller APIs? And a separate one for decommission, which is a client API. 

Also, it would be nice to have JIRAs since these are significant additions. Thanks!

**Comment by aloknnikhil** [comment]
@hachikuji Fixed it up to merge only the Broker-Controller RPCs (Heartbeat/Registration)

File: clients/src/main/java/org/apache/kafka/common/protocol/ApiKeys.java [hunk 1]
```diff
@@ -98,7 +98,9 @@
     UPDATE_FEATURES(ApiMessageType.UPDATE_FEATURES, false, true),
     ENVELOPE(ApiMessageType.ENVELOPE, true, RecordBatch.MAGIC_VALUE_V0, false, true),
     FETCH_SNAPSHOT(ApiMessageType.FETCH_SNAPSHOT, false, RecordBatch.MAGIC_VALUE_V0, false, true),
-    DESCRIBE_CLUSTER(ApiMessageType.DESCRIBE_CLUSTER);
+    DESCRIBE_CLUSTER(ApiMessageType.DESCRIBE_CLUSTER),
+    BROKER_REGISTRATION(ApiMessageType.BROKER_REGISTRATION, false, RecordBatch.MAGIC_VALUE_V0, false, true),
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/common/protocol/ApiKeys.java]: I think the second flag should be true for both of these. They do require CLUSTER_ACTION.

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 2]
```diff
@@ -232,6 +232,9 @@ class KafkaApis(val requestChannel: RequestChannel,
         case ApiKeys.DESCRIBE_QUORUM => requestHelper.closeConnection(request, util.Collections.emptyMap())
         case ApiKeys.FETCH_SNAPSHOT => requestHelper.closeConnection(request, util.Collections.emptyMap())
 
+        // Handle requests that should have been sent to the KIP-500 controller.
```
Comment (hachikuji) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: nit: this applies to all of the Raft apis above as well

**Review by hachikuji — COMMENTED** [review]
Left a couple comments. You will need to update `RequestConvertToJson` as well. Probably we should have some tests in `RequestResponseTest` as well.

File: clients/src/main/java/org/apache/kafka/common/protocol/ApiKeys.java (hunk 1 above)
Comment (aloknnikhil) [reply clients/src/main/java/org/apache/kafka/common/protocol/ApiKeys.java]: Ack.

File: core/src/main/scala/kafka/server/KafkaApis.scala (hunk 2 above)
Comment (aloknnikhil) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: Ack.

**Comment by aloknnikhil** [comment]
> Left a couple comments. You will need to update RequestConvertToJson as well. Probably we should have some tests in RequestResponseTest as well.

Done.

**Review by hachikuji — APPROVED** [review]
LGTM. Thanks for the patch!

**Comment by hachikuji** [comment]
@aloknnikhil Looks like we will need to fix some failures in `ApiKeysTest` before we can merge this. I suspect that the test does not like the throttle time fields. One way we could handle this is to ignore `controllerOnly` APIs in this test case. I guess we haven't really figured out the throttling picture for broker-to-controller APIs.

**Pull request merged**